### PR TITLE
[Merged by Bors] - Implement `is_identifier_(start/part)` using `icu_properties`

### PR DESCRIPTION
**Pull request by jedel1043**
As mentioned in https://github.com/boa-dev/boa/pull/2848#issuecomment-1518909512, this uses our new default ICU4X data to replace `char::is_start` and `char::is_continue` from the `boa_unicode` crate with the [`icu_properties`](https://crates.io/crates/icu_properties) crate.

Note that this doesn't deprecate `boa_unicode` yet, since that'll require some discussion about how to proceed with a now unused sub-crate.

### CODE REVIEW

File: boa_parser/src/lexer/identifier.rs [hunk 1]
```diff
@@ -6,9 +6,31 @@ use crate::lexer::{
 use boa_ast::{Keyword, Position, Span};
 use boa_interner::Interner;
 use boa_profiler::Profiler;
-use boa_unicode::UnicodeProperties;
+use icu_properties::sets::{CodePointSetData, CodePointSetDataBorrowed};
+use once_cell::sync::Lazy;
 use std::io::Read;
 
+struct PropertySets {
```
Comment (Razican) [inline boa_parser/src/lexer/identifier.rs:13]: Could we add some documentation here?

**Review by Razican — APPROVED** [review]
Looks good to me! Regarding `boa_unicodde`, we could maybe use `cargo-tally` (https://github.com/dtolnay/cargo-tally).

With this we could check if someone else uses it and deprecate our crate.

**Comment by jedel1043** [comment]
> Looks good to me! Regarding `boa_unicodde`, we could maybe use `cargo-tally` (https://github.com/dtolnay/cargo-tally).
> 
> With this we could check if someone else uses it and deprecate our crate.

Apparently there are no dependents of `boa_unicode` apart from `Boa` and `boa_engine` (https://crates.io/crates/boa_unicode/reverse_dependencies), so I think it should be safe to deprecate it.

**Comment by Razican** [comment]
> > Looks good to me! Regarding `boa_unicodde`, we could maybe use `cargo-tally` (https://github.com/dtolnay/cargo-tally).
> > With this we could check if someone else uses it and deprecate our crate.
> 
> Apparently there are no dependents of `boa_unicode` apart from `Boa` and `boa_engine` (https://crates.io/crates/boa_unicode/reverse_dependencies), so I think it should be safe to deprecate it.

Let's do it! :)

**Review by HalidOdat — APPROVED** [review]
Looks good to me!

**Comment by jedel1043** [comment]
Then, after this gets merged, we can open a new branch to bump `icu_unicode` to 0.16.1 with a description signaling its deprecated state, and delete the crate from our main branch.

**Comment by raskad** [comment]
For the sake of documentation; I noticed that after this change our benchmarks for the parser looked better so I ran them and compared cebec9da909d7a51308306490d8b25e809f0babc to c33000591252c398dde5997289709d8989645afe.
This change seems to have improved the performance of the parser by a good amount:

|Benchmark|Changes|
|---|---|
|Symbols (Parser)                        | [-13.850% -12.097% -10.548%]|
|For loop (Parser)                       | [-5.3454% -4.3440% -3.4117%]|
|Fibonacci (Parser)                      | [-13.609% -10.961% -8.5178%]|
|Object Creation (Parser)                | [-8.1511% -6.9890% -5.7952%]|
|Static Object Property Access (Parser)  | [-10.387% -9.2301% -8.0496%]|
|Dynamic Object Property Access (Parser) | [-7.9534% -6.8847% -5.7769%]|
|RegExp Literal Creation (Parser)        | [-7.4906% -6.3700% -5.4132%]|
|RegExp Creation (Parser)                | [-10.154% -9.1298% -7.9548%]|
|RegExp Literal (Parser)                 | [-7.7712% -6.7077% -5.5830%]|
|RegExp (Parser)                         | [-12.477% -11.538% -10.613%]|
|Array access (Parser)                   | [-15.038% -13.548% -11.900%]|
|Array creation (Parser)                 | [-17.589% -15.650% -13.799%]|
|Array pop (Parser)                      | [-8.6277% -7.0937% -5.5033%]|
|String concatenation (Parser)           | [-12.413% -10.984% -9.6802%]|
|String comparison (Parser)              | [-8.1121% -6.9242% -5.6435%]|
|String copy (Parser)                    | [-8.6093% -7.1169% -5.6620%]|
|Number Object Access (Parser)           | [-10.555% -9.5270% -8.4492%]|
|Boolean Object Access (Parser)          | [-17.107% -16.002% -14.850%]|
|String Object Access (Parser)           | [-16.571% -15.397% -14.316%]|
|Arithmetic operations (Parser)          | [+1.5678% +3.3078% +5.0827%]|
|Clean js (Parser)                       | [-7.3581% -6.1867% -4.6942%]|
|Mini js (Parser)                        | [-6.1803% -5.2065% -4.2830%]|

cc @jedel1043

**Comment by jedel1043** [comment]
Nice! And it makes much sense, since `icu_properties` is hyperoptimized for char queries, whereas our old implementation was just a succession of queries using several tables and property matches that made the categorization work a lot slower.

**Pull request closed without merging**